### ISSUE TITLE
mpiexec --cpu-set  ORTE error 

### ISSUE THREAD

**Issue by GuillaumeMercier** [open-mpi/ompi#5020]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
Git repo version: v2.x-dev-5428-g17a5416

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
from a git clone (v2.x-dev-5428-g17a5416)

### Please describe the system on which you are running

* Operating system/version:  3.10.0-327.36.3.el7.x86_64
* Computer hardware:  Xeon E5 V3
* Network type: OPA

-----------------------------

## Details of the problem

Please describe, in detail, the problem that you are having, including the behavior you expect to see, the actual behavior that you are seeing, steps to reproduce the problem, etc.  It is most helpful if you can attach a small program that a developer can use to reproduce your problem.

When trying to launch a programm   with the `--cpu-set` option, I've got this:

```shell
[miriel016:115300] [[3216,0],0] ORTE_ERROR_LOG: Data unpack would read past end of buffer in file ../../../../ompi-sources/orte/mca/odls/base/odls_base_default_fns.c at line 565
```

My command line is:
```shell
mpiexec --cpu-set 0,1,2,12,13,14 --bind-to core  -np 4 -host localhost sample_prog   4
```

**Comment by ggouaillardet** [comment]
note `Git repo version` is misleading here (a known git thing ...), and the issue occurs on the (latest) `master` branch.

can you please give the inline patch a try ?

```patch
diff --git a/orte/mca/odls/base/odls_base_default_fns.c b/orte/mca/odls/base/odls_base_default_fns.c
index ff9426b..be20292 100644
--- a/orte/mca/odls/base/odls_base_default_fns.c
+++ b/orte/mca/odls/base/odls_base_default_fns.c
@@ -15,7 +15,7 @@
  *                         All rights reserved.
  * Copyright (c) 2011-2017 Cisco Systems, Inc.  All rights reserved
  * Copyright (c) 2013-2018 Intel, Inc.  All rights reserved.
- * Copyright (c) 2014-2017 Research Organization for Information Science
+ * Copyright (c) 2014-2018 Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
  * Copyright (c) 2017      Mellanox Technologies Ltd. All rights reserved.
  * Copyright (c) 2017      IBM Corporation. All rights reserved.
@@ -178,7 +178,11 @@ int orte_odls_base_default_get_add_procs_data(opal_buffer_t *buffer,
     } else {
         nidmap = NULL;
     }
-    opal_dss.pack(buffer, &nidmap, 1, OPAL_STRING);
+    if (ORTE_SUCCESS != (rc = opal_dss.pack(buffer, &nidmap, 1, OPAL_STRING))) {
+        ORTE_ERROR_LOG(rc);
+        free(nidmap);
+        return rc;
+    }
     if (NULL != nidmap) {
         free(nidmap);
     }
@@ -375,11 +379,15 @@ int orte_odls_base_default_get_add_procs_data(opal_buffer_t *buffer,
             ORTE_ERROR_LOG(rc);
             return rc;
         }
-        if (ORTE_SUCCESS != (rc = opal_dss.pack(buffer, &nidmap, 1, OPAL_STRING))) {
-            ORTE_ERROR_LOG(rc);
-            free(nidmap);
-            return rc;
-        }
+    } else {
+        nidmap = NULL;
+    }
+    if (ORTE_SUCCESS != (rc = opal_dss.pack(buffer, &nidmap, 1, OPAL_STRING))) {
+        ORTE_ERROR_LOG(rc);
+        free(nidmap);
+        return rc;
+    }
+    if (NULL != nidmap) {
         free(nidmap);
     }
 
```

@rhc54 would you mind commenting on it if/when you get a chance ?

that seems good enough, but it could be suboptimal.

**Comment by ggouaillardet** [comment]
ok, i will try again on Monday.

meanwhile, can you post the content of your `machinefile` and the (text) output of `lstopo` ?

fwiw, I did
```
mpirun --cpu-set ... --bind-to core grep Cpus_allowed_list /proc/self/status
```
and as far as I remember, processes were (correctly) bound to distinct cores

**Comment by rhc54** [comment]
FWIW: I'm pretty sure we won't bind if you set --oversubscribe, regardless of the rest of the command line.

**Comment by GuillaumeMercier** [comment]
Ok, I removed this option and the result is the same.

**Comment by rhc54** [comment]
Relates to #5266

**Comment by rhc54** [comment]
Looks like the ```--cpu-set``` option is broken

**Comment by jithinjosepkl** [comment]
I'm also getting this error with OpenMPI 4.0.0 with the --cpu-set option. Is there any alternative to --cpu-set option?

**Comment by markalle** [comment]
Fwiw I think the below PR fixes this issue:
https://github.com/open-mpi/ompi/pull/6584
I was working on a slightly different use-case there, I think my fix makes the command from this issue pass as well.

**Comment by hppritcha** [comment]
I can't reproduce using the Open MPI 4.0.6rc2.   Closing.  Please reopen if this isn't fixed for you in the 4.0.6 or later release.
